File: docs/competitors/c++/Eratosthenes/Eratosthenes.cpp
#include <iostream>
#include <vector>

std::vector<int> generatePrimes(int maxNumber)
{
	std::vector<int> primes;

	if (maxNumber < 2)
	{
		return primes;
	}

	std::vector<bool> isPrime(maxNumber + 1, true);
	isPrime[0] = isPrime[1] = false;

	for (int currentNumber = 2; currentNumber * currentNumber <= maxNumber; ++currentNumber)
	{
		if (isPrime[currentNumber])
		{
			for (int multiple = currentNumber * currentNumber; multiple <= maxNumber; multiple += currentNumber)
			{
				isPrime[multiple] = false;
			}
		}
	}

	for (int number = 2; number <= maxNumber; ++number)
	{
		if (isPrime[number])
		{
			primes.push_back(number);
		}
	}

	return primes;
}

int main()
{
	int userInput;
	std::cout << "Enter an integer: ";
	std::cin >> userInput;

	std::vector<int> primes = generatePrimes(userInput);

	if (primes.empty())
	{
		std::cout << "No prime numbers" << std::endl;
	}
	else
	{
		std::cout << "Prime numbers: ";
		for (int prime : primes)
		{
			std::cout << prime << " ";
		}
		std::cout << std::endl;
	}

	return 0;
}

File: docs/competitors/pascal/Eratosthenes/Eratosthenes.pas
program Eratosthenes;

function GeneratePrimes(maxNumber: integer): array of integer;
var
  isPrime: array of boolean;
  count: integer;
  i, j: integer;
begin
  if maxNumber < 2 then
  begin
    SetLength(Result, 0);
    Exit;
  end;

  SetLength(isPrime, maxNumber + 1);
  for var k := 0 to maxNumber do
    isPrime[k] := true;

  isPrime[0] := false;
  isPrime[1] := false;

  i := 2;
  while i * i <= maxNumber do
  begin
    if isPrime[i] then
    begin
      j := i * i;
      while j <= maxNumber do
      begin
        isPrime[j] := false;
        j := j + i;
      end;
    end;
    i := i + 1;
  end;

  count := 0;
  for var k := 2 to maxNumber do
    if isPrime[k] then
      count := count + 1;

  SetLength(Result, count);
  count := 0;
  for var k := 2 to maxNumber do
    if isPrime[k] then
    begin
      Result[count] := k;
      count := count + 1;
    end;
end;

var
  n: integer;
  primes: array of integer;
begin
  write('Enter an integer: ');
  readln(n);
  
  primes := GeneratePrimes(n);
  
  if Length(primes) = 0 then
    writeln('No prime numbers')
  else
  begin
    write('Prime numbers: ');
    for var i := 0 to Length(primes) - 1 do
      write(primes[i], ' ');
    writeln;
  end;
end.